File: pypp_cli/test_dir/cpp/src/interfaces/with_dataclasses.cpp
#include "src/interfaces/with_dataclasses.h"
#include "pypp_util/print.h"

namespace me {
void GreeterImpl::greet(int a) {
    pypp::print(pypp::PyStr(std::format("{} says hello {} times!", name, a)));
}

void interface_with_dataclasses_fn() {
    pypp::print(pypp::PyStr("INTERFACE WITH DATACLASSES RESULTS:"));
    GreeterImpl g = GreeterImpl(pypp::PyStr("Alice"));
    g.greet(3);
}

} // namespace me

File: pypp_cli/test_dir/src/interfaces/with_dataclasses.py
from abc import ABC, abstractmethod
from pypp_python import Valu, dataclass


class GreeterInterface(ABC):
    @abstractmethod
    def greet(self, a: int):
        pass


@dataclass
class GreeterImpl(GreeterInterface):
    name: Valu(str)

    def greet(self, a: int):
        print(f"{self.name} says hello {a} times!")


def interface_with_dataclasses_fn():
    print("INTERFACE WITH DATACLASSES RESULTS:")
    g: GreeterImpl = GreeterImpl("Alice")
    g.greet(3)
